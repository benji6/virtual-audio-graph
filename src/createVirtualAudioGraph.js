/* global AudioContext */
import {equals, forEach} from './utils'
import connectAudioNodes from './connectAudioNodes'
import createVirtualAudioNode from './createVirtualAudioNode'

const disconnectParents = (virtualNode, virtualNodes) => forEach(
  key => virtualNodes[key].disconnect(virtualNode),
  Object.keys(virtualNodes)
)

export default ({
  audioContext = new AudioContext(),
  output = audioContext.destination,
} = {}) => {
  return {
    audioContext,
    get currentTime () { return audioContext.currentTime },
    getAudioNodeById (id) { return this.virtualNodes[id].audioNode },
    update (newGraph) {
      forEach(id => {
        if (newGraph.hasOwnProperty(id)) return
        const virtualAudioNode = this.virtualNodes[id]
        virtualAudioNode.disconnectAndDestroy()
        disconnectParents(virtualAudioNode, this.virtualNodes)
        delete this.virtualNodes[id]
      }, Object.keys(this.virtualNodes))

      forEach(key => {
        if (key === 'output') throw new Error('"output" is not a valid id')
        const newNodeParams = newGraph[key]
        const [paramsNodeName, paramsOutput, paramsParams] = newNodeParams
        const virtualAudioNode = this.virtualNodes[key]
        if (virtualAudioNode == null) {
          this.virtualNodes[key] = createVirtualAudioNode(audioContext, newNodeParams)
          return
        }
        if (
          (paramsParams && paramsParams.startTime) !==
            (virtualAudioNode.params && virtualAudioNode.params.startTime) ||
          (paramsParams && paramsParams.stopTime) !==
            (virtualAudioNode.params && virtualAudioNode.params.stopTime) ||
          paramsNodeName !== virtualAudioNode.node
        ) {
          virtualAudioNode.disconnectAndDestroy()
          disconnectParents(virtualAudioNode, this.virtualNodes)
          this.virtualNodes[key] = createVirtualAudioNode(audioContext, newNodeParams)
          return
        }
        if (!equals(paramsOutput, virtualAudioNode.output)) {
          virtualAudioNode.disconnect()
          disconnectParents(virtualAudioNode, this.virtualNodes)
          virtualAudioNode.output = paramsOutput
        }

        virtualAudioNode.update(paramsParams)
      }, Object.keys(newGraph))

      connectAudioNodes(
        this.virtualNodes,
        virtualNode => virtualNode.connect(output)
      )

      return this
    },
    virtualNodes: {},
  }
}
