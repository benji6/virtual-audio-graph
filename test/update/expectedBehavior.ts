import createVirtualAudioGraph, * as V from "../../src";
import pingPongDelay from "../utils/pingPongDelay";
import sineOsc from "../utils/sineOsc";

const audioContext: any = new AudioContext();
const virtualAudioGraph = createVirtualAudioGraph({ audioContext });

describe("expected behavior with update", () => {
  test("returns itself", () => {
    const virtualNodeParams = { 0: V.oscillator("output", { type: "square" }) };

    expect(virtualAudioGraph.update(virtualNodeParams)).toBe(virtualAudioGraph);
  });

  test('throws error if node ID is "output"', () => {
    expect(() =>
      virtualAudioGraph.update({
        output: V.oscillator("output"),
      }),
    ).toThrow(
      '"output" is a virtual-audio-graph reserved string and therefore not a valid node ID',
    );
  });

  test("throws error if output string id is null", () => {
    expect(() =>
      virtualAudioGraph.update({
        bar: V.oscillator(null as any),
        foo: V.gain("output"),
      }),
    ).toThrow("Output not specified for oscillator");
  });

  test("throws error if output object key is null", () => {
    expect(() =>
      virtualAudioGraph.update({
        bar: V.oscillator({ key: null } as any),
        foo: V.gain("output"),
      }),
    ).toThrow('node with ID "bar" does not specify an output');
  });

  test("throws error if output string id is not a node ID", () => {
    expect(() =>
      virtualAudioGraph.update({
        bar: V.oscillator("baz"),
        foo: V.gain("output"),
      }),
    ).toThrow(
      'node with ID "bar" specifies an output ID "baz", but no such node exists',
    );
  });

  test("throws error if output object key is not a node ID", () => {
    expect(() =>
      virtualAudioGraph.update({
        bar: V.oscillator({ key: "baz" }),
        foo: V.gain("output"),
      }),
    ).toThrow(
      'node with ID "bar" specifies an output ID "baz", but no such node exists',
    );
  });

  test("adds then removes nodes", () => {
    virtualAudioGraph.update({
      0: V.gain("output"),
      1: V.oscillator(0),
    });

    expect(audioContext.toJSON()).toMatchSnapshot();

    virtualAudioGraph.update({});

    expect(audioContext.toJSON()).toMatchSnapshot();
  });

  test("handles random strings for ids", () => {
    virtualAudioGraph.update({
      bar: V.oscillator("foo"),
      foo: V.gain("output"),
    });
    expect(audioContext.toJSON()).toMatchSnapshot();
  });

  test("handles simple updates", () => {
    virtualAudioGraph.update({});
    expect(audioContext.toJSON()).toMatchSnapshot();

    virtualAudioGraph.update({ 0: V.gain("output") });
    expect(audioContext.toJSON()).toMatchSnapshot();

    virtualAudioGraph.update({ 0: V.oscillator("output") });
    expect(audioContext.toJSON()).toMatchSnapshot();

    virtualAudioGraph.update({ 0: pingPongDelay("output") });
    expect(audioContext.toJSON()).toMatchSnapshot();
  });

  test("updates standard and custom nodes if passed same id but different params", () => {
    virtualAudioGraph.update({
      0: V.oscillator("output", { detune: -9, frequency: 220 }),
    });

    expect(audioContext.toJSON()).toMatchSnapshot();

    virtualAudioGraph.update({
      0: V.oscillator("output", { detune: 0, frequency: 880 }),
    });

    expect(audioContext.toJSON()).toMatchSnapshot();

    virtualAudioGraph.update({
      0: sineOsc("output", { frequency: 110, gain: 0.5 }),
    });

    expect(audioContext.toJSON()).toMatchSnapshot();

    virtualAudioGraph.update({
      0: sineOsc("output", { frequency: 660, gain: 0.2 }),
    });

    expect(audioContext.toJSON()).toMatchSnapshot();
  });

  test("connects nodes to each other", () => {
    virtualAudioGraph.update({
      0: V.gain("output"),
      1: V.oscillator(0),
    });
    expect(audioContext.toJSON()).toMatchSnapshot();

    virtualAudioGraph.update({
      0: V.oscillator("output"),
    });

    expect(audioContext.toJSON()).toMatchSnapshot();
  });

  test("reconnects nodes to each other", () => {
    virtualAudioGraph.update({
      0: V.gain("output"),
      1: V.oscillator(0),
    });
    expect(audioContext.toJSON()).toMatchSnapshot();

    virtualAudioGraph.update({
      0: V.gain("output"),
      1: V.oscillator("output"),
    });

    expect(audioContext.toJSON()).toMatchSnapshot();
  });

  test("connects and reconnects nodes to audioParams", () => {
    virtualAudioGraph.update({
      0: V.gain("output"),
      1: V.oscillator(0),
      2: V.oscillator(
        { destination: "frequency", key: 1 },
        { frequency: 0.5, type: "triangle" },
      ),
    });

    expect(audioContext.toJSON()).toMatchSnapshot();

    virtualAudioGraph.update({
      0: V.gain("output"),
      1: V.oscillator(0),
      2: V.oscillator([{ destination: "detune", key: 1 }], {
        frequency: 0.5,
        type: "triangle",
      }),
    });

    expect(audioContext.toJSON()).toMatchSnapshot();

    virtualAudioGraph.update({
      0: V.oscillator("output"),
    });

    expect(audioContext.toJSON()).toMatchSnapshot();
  });

  test("disconnects and reconnects child nodes properly", () => {
    virtualAudioGraph.update({
      0: V.gain("output"),
      1: V.stereoPanner(0),
      2: V.gain(1),
    });

    expect(audioContext.toJSON()).toMatchSnapshot();

    virtualAudioGraph.update({
      0: V.gain("output"),
      1: V.gain(0),
      2: V.gain(1),
    });

    expect(audioContext.toJSON()).toMatchSnapshot();
  });
});
