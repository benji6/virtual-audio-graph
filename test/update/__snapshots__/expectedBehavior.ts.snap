// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expected behavior with update adds then removes nodes 1`] = `
{
  "inputs": [
    {
      "gain": {
        "inputs": [],
        "value": 1,
      },
      "inputs": [
        {
          "detune": {
            "inputs": [],
            "value": 0,
          },
          "frequency": {
            "inputs": [],
            "value": 440,
          },
          "inputs": [],
          "name": "OscillatorNode",
          "type": "sine",
        },
      ],
      "name": "GainNode",
    },
  ],
  "name": "AudioDestinationNode",
}
`;

exports[`expected behavior with update adds then removes nodes 2`] = `
{
  "inputs": [],
  "name": "AudioDestinationNode",
}
`;

exports[`expected behavior with update connects and reconnects nodes to audioParams 1`] = `
{
  "inputs": [
    {
      "gain": {
        "inputs": [],
        "value": 1,
      },
      "inputs": [
        {
          "detune": {
            "inputs": [],
            "value": 0,
          },
          "frequency": {
            "inputs": [
              {
                "detune": {
                  "inputs": [],
                  "value": 0,
                },
                "frequency": {
                  "inputs": [],
                  "value": 0.5,
                },
                "inputs": [],
                "name": "OscillatorNode",
                "type": "triangle",
              },
            ],
            "value": 440,
          },
          "inputs": [],
          "name": "OscillatorNode",
          "type": "sine",
        },
      ],
      "name": "GainNode",
    },
  ],
  "name": "AudioDestinationNode",
}
`;

exports[`expected behavior with update connects and reconnects nodes to audioParams 2`] = `
{
  "inputs": [
    {
      "gain": {
        "inputs": [],
        "value": 1,
      },
      "inputs": [
        {
          "detune": {
            "inputs": [
              {
                "detune": {
                  "inputs": [],
                  "value": 0,
                },
                "frequency": {
                  "inputs": [],
                  "value": 0.5,
                },
                "inputs": [],
                "name": "OscillatorNode",
                "type": "triangle",
              },
            ],
            "value": 0,
          },
          "frequency": {
            "inputs": [],
            "value": 440,
          },
          "inputs": [],
          "name": "OscillatorNode",
          "type": "sine",
        },
      ],
      "name": "GainNode",
    },
  ],
  "name": "AudioDestinationNode",
}
`;

exports[`expected behavior with update connects and reconnects nodes to audioParams 3`] = `
{
  "inputs": [
    {
      "detune": {
        "inputs": [],
        "value": 0,
      },
      "frequency": {
        "inputs": [],
        "value": 440,
      },
      "inputs": [],
      "name": "OscillatorNode",
      "type": "sine",
    },
  ],
  "name": "AudioDestinationNode",
}
`;

exports[`expected behavior with update connects nodes to each other 1`] = `
{
  "inputs": [
    {
      "gain": {
        "inputs": [],
        "value": 1,
      },
      "inputs": [
        {
          "detune": {
            "inputs": [],
            "value": 0,
          },
          "frequency": {
            "inputs": [],
            "value": 440,
          },
          "inputs": [],
          "name": "OscillatorNode",
          "type": "sine",
        },
      ],
      "name": "GainNode",
    },
  ],
  "name": "AudioDestinationNode",
}
`;

exports[`expected behavior with update connects nodes to each other 2`] = `
{
  "inputs": [
    {
      "detune": {
        "inputs": [],
        "value": 0,
      },
      "frequency": {
        "inputs": [],
        "value": 440,
      },
      "inputs": [],
      "name": "OscillatorNode",
      "type": "sine",
    },
  ],
  "name": "AudioDestinationNode",
}
`;

exports[`expected behavior with update disconnects and reconnects child nodes properly 1`] = `
{
  "inputs": [
    {
      "gain": {
        "inputs": [],
        "value": 1,
      },
      "inputs": [
        {
          "inputs": [
            {
              "gain": {
                "inputs": [],
                "value": 1,
              },
              "inputs": [],
              "name": "GainNode",
            },
          ],
          "name": "StereoPannerNode",
          "pan": {
            "inputs": [],
            "value": 0,
          },
        },
      ],
      "name": "GainNode",
    },
  ],
  "name": "AudioDestinationNode",
}
`;

exports[`expected behavior with update disconnects and reconnects child nodes properly 2`] = `
{
  "inputs": [
    {
      "gain": {
        "inputs": [],
        "value": 1,
      },
      "inputs": [
        {
          "gain": {
            "inputs": [],
            "value": 1,
          },
          "inputs": [
            {
              "gain": {
                "inputs": [],
                "value": 1,
              },
              "inputs": [],
              "name": "GainNode",
            },
          ],
          "name": "GainNode",
        },
      ],
      "name": "GainNode",
    },
  ],
  "name": "AudioDestinationNode",
}
`;

exports[`expected behavior with update handles random strings for ids 1`] = `
{
  "inputs": [
    {
      "gain": {
        "inputs": [],
        "value": 1,
      },
      "inputs": [
        {
          "detune": {
            "inputs": [],
            "value": 0,
          },
          "frequency": {
            "inputs": [],
            "value": 440,
          },
          "inputs": [],
          "name": "OscillatorNode",
          "type": "sine",
        },
      ],
      "name": "GainNode",
    },
  ],
  "name": "AudioDestinationNode",
}
`;

exports[`expected behavior with update handles random strings for ids 2`] = `
{
  "inputs": [],
  "name": "AudioDestinationNode",
}
`;

exports[`expected behavior with update handles random strings for ids 3`] = `
{
  "inputs": [
    {
      "gain": {
        "inputs": [],
        "value": 1,
      },
      "inputs": [],
      "name": "GainNode",
    },
  ],
  "name": "AudioDestinationNode",
}
`;

exports[`expected behavior with update handles random strings for ids 4`] = `
{
  "inputs": [
    {
      "detune": {
        "inputs": [],
        "value": 0,
      },
      "frequency": {
        "inputs": [],
        "value": 440,
      },
      "inputs": [],
      "name": "OscillatorNode",
      "type": "sine",
    },
  ],
  "name": "AudioDestinationNode",
}
`;

exports[`expected behavior with update handles random strings for ids 5`] = `
{
  "inputs": [
    {
      "inputs": [
        {
          "delayTime": {
            "inputs": [],
            "value": 0.3333333333333333,
          },
          "inputs": [
            {
              "gain": {
                "inputs": [],
                "value": 0.3333333333333333,
              },
              "inputs": [
                {
                  "delayTime": {
                    "inputs": [],
                    "value": 0.3333333333333333,
                  },
                  "inputs": [
                    {
                      "gain": {
                        "inputs": [],
                        "value": 0.3333333333333333,
                      },
                      "inputs": [
                        "<circular:DelayNode>",
                      ],
                      "name": "GainNode",
                    },
                  ],
                  "name": "DelayNode",
                },
              ],
              "name": "GainNode",
            },
          ],
          "name": "DelayNode",
        },
      ],
      "name": "StereoPannerNode",
      "pan": {
        "inputs": [],
        "value": 1,
      },
    },
    {
      "inputs": [
        {
          "delayTime": {
            "inputs": [],
            "value": 0.3333333333333333,
          },
          "inputs": [
            {
              "gain": {
                "inputs": [],
                "value": 0.3333333333333333,
              },
              "inputs": [
                {
                  "delayTime": {
                    "inputs": [],
                    "value": 0.3333333333333333,
                  },
                  "inputs": [
                    {
                      "gain": {
                        "inputs": [],
                        "value": 0.3333333333333333,
                      },
                      "inputs": [
                        "<circular:DelayNode>",
                      ],
                      "name": "GainNode",
                    },
                  ],
                  "name": "DelayNode",
                },
              ],
              "name": "GainNode",
            },
          ],
          "name": "DelayNode",
        },
      ],
      "name": "StereoPannerNode",
      "pan": {
        "inputs": [],
        "value": -1,
      },
    },
  ],
  "name": "AudioDestinationNode",
}
`;

exports[`expected behavior with update reconnects nodes to each other 1`] = `
{
  "inputs": [
    {
      "gain": {
        "inputs": [],
        "value": 1,
      },
      "inputs": [
        {
          "detune": {
            "inputs": [],
            "value": 0,
          },
          "frequency": {
            "inputs": [],
            "value": 440,
          },
          "inputs": [],
          "name": "OscillatorNode",
          "type": "sine",
        },
      ],
      "name": "GainNode",
    },
  ],
  "name": "AudioDestinationNode",
}
`;

exports[`expected behavior with update reconnects nodes to each other 2`] = `
{
  "inputs": [
    {
      "gain": {
        "inputs": [],
        "value": 1,
      },
      "inputs": [],
      "name": "GainNode",
    },
    {
      "detune": {
        "inputs": [],
        "value": 0,
      },
      "frequency": {
        "inputs": [],
        "value": 440,
      },
      "inputs": [],
      "name": "OscillatorNode",
      "type": "sine",
    },
  ],
  "name": "AudioDestinationNode",
}
`;

exports[`expected behavior with update updates standard and custom nodes if passed same id but different params 1`] = `
{
  "inputs": [
    {
      "detune": {
        "inputs": [],
        "value": -9,
      },
      "frequency": {
        "inputs": [],
        "value": 220,
      },
      "inputs": [],
      "name": "OscillatorNode",
      "type": "sine",
    },
  ],
  "name": "AudioDestinationNode",
}
`;

exports[`expected behavior with update updates standard and custom nodes if passed same id but different params 2`] = `
{
  "inputs": [
    {
      "detune": {
        "inputs": [],
        "value": 0,
      },
      "frequency": {
        "inputs": [],
        "value": 880,
      },
      "inputs": [],
      "name": "OscillatorNode",
      "type": "sine",
    },
  ],
  "name": "AudioDestinationNode",
}
`;

exports[`expected behavior with update updates standard and custom nodes if passed same id but different params 3`] = `
{
  "inputs": [
    {
      "gain": {
        "inputs": [],
        "value": 0.5,
      },
      "inputs": [
        {
          "detune": {
            "inputs": [],
            "value": 0,
          },
          "frequency": {
            "inputs": [],
            "value": 110,
          },
          "inputs": [],
          "name": "OscillatorNode",
          "type": "sine",
        },
      ],
      "name": "GainNode",
    },
  ],
  "name": "AudioDestinationNode",
}
`;

exports[`expected behavior with update updates standard and custom nodes if passed same id but different params 4`] = `
{
  "inputs": [
    {
      "gain": {
        "inputs": [],
        "value": 0.2,
      },
      "inputs": [
        {
          "detune": {
            "inputs": [],
            "value": 0,
          },
          "frequency": {
            "inputs": [],
            "value": 660,
          },
          "inputs": [],
          "name": "OscillatorNode",
          "type": "sine",
        },
      ],
      "name": "GainNode",
    },
  ],
  "name": "AudioDestinationNode",
}
`;
